kview/src/main/java/dev/ikm/komet/kview/mvvm/view/login/LoginPageController.java

```diff
@@ -0,0 +1,132 @@
+     *
+     * @param location  The location used to resolve relative paths for the root object, or {@code null} if not known.
+     * @param resources The resources used to localize the root object, or {@code null} if not localized.
+     */
+    @Override
+    public void initialize(URL location, ResourceBundle resources) {
+        ChangeListener<String> isNotPopulatedListener = (observable, oldValue, newValue) ->
+                loginViewModel.getValidators(IS_NOT_POPULATED).stream()
+                .findAny()
+                .ifPresent(validator -> validator.apply(null, loginViewModel));
+
+        usernameTextField.textProperty().addListener(isNotPopulatedListener);
+        passwordField.textProperty().addListener(isNotPopulatedListener);
+
+        usernameTextField.textProperty().bindBidirectional(loginViewModel.getProperty(USERNAME));
+        passwordField.textProperty().bindBidirectional(loginViewModel.getProperty(PASSWORD));
+        signInButton.disableProperty().bind(loginViewModel.getProperty(SIGN_IN_BUTTON_STATE));
+        authErrorLabel.textProperty().bind(loginViewModel.getProperty(AUTH_ERROR));
+    }
+
+    /**
+     * Handles the action event when the sign-in button is clicked.
+     *
+     * @param actionEvent The action event triggered by the sign-in button.
+     */
+    @FXML
+    private void signInAction(ActionEvent actionEvent) {
+        loginViewModel.save();
+
+        if (loginViewModel.hasErrorMsgs()) {
+            String usernameErrorMessage = "";
+            String passwordErrorMessage = "";
+
+            for (ValidationMessage validationMessage : loginViewModel.getValidationMessages()) {
+                loginViewModel.updateErrors(validationMessage);
+
+                String propName = validationMessage.propertyName();
+                String message = validationMessage.interpolate(loginViewModel);
+
+                if (USERNAME.name().equals(propName)) {
+                    usernameErrorMessage = message;
+                } else if (PASSWORD.name().equals(propName)) {
+                    passwordErrorMessage = message;
+                }
+            }
+
+            // set the error labels after processing all validation messages
+            usernameErrorLabel.setText(usernameErrorMessage);
+            passwordErrorLabel.setText(passwordErrorMessage);
+        } else {
+            // Perform authentication
+            final String username = loginViewModel.getValue(USERNAME);
+            final String password = loginViewModel.getValue(PASSWORD);
+            loginViewModel.authenticate(username, password);
+
+            // Clear the error labels
+            usernameErrorLabel.setText("");
+            passwordErrorLabel.setText("");
+
+            // Clear view model's values to remove passwords from memory
+            loginViewModel.setValue(USERNAME, "");
+            loginViewModel.setValue(PASSWORD, "");
+            loginViewModel.save(true);
```

Comment:
Thanks for the detailed explanation, Carl! Using the `reset()` method is a better solution. However, as you mentioned, it also clears the values from the screen, which isn’t ideal when, for example, a wrong password is entered, we don’t want to clear the text fields in that case, forcing the user to re-enter their credentials. The good part is that TestFX unit test caught this behaviour change, which is great :).
To address [this](https://github.com/ikmdev/komet/pull/93/commits/f6e877409b892b93851c9796a72844e1fc6e26e9), I moved the same block into the Login view model so that the text fields are only cleared after successful authentication.